test/Microsoft.AspNetCore.Mvc.Razor.ViewCompilation.FunctionalTests/Infrastructure/ApplicationTestFixture.cs

```diff
@@ -5,66 +5,57 @@
 using System.Collections.Generic;
 using System.IO;
 using System.Net.Http;
-using System.Runtime.InteropServices;
 using Microsoft.AspNetCore.Server.IntegrationTesting;
-using Microsoft.DotNet.Cli.Utils;
 using Microsoft.Extensions.Logging;
 
 namespace Microsoft.AspNetCore.Mvc.Razor.ViewCompilation
 {
     public abstract class ApplicationTestFixture : IDisposable
     {
-        public const string NuGetPackagesEnvironmentKey = "NUGET_PACKAGES";
-        public const string DotnetSkipFirstTimeExperience = "DOTNET_SKIP_FIRST_TIME_EXPERIENCE";
         public const string DotnetCLITelemetryOptOut = "DOTNET_CLI_TELEMETRY_OPTOUT";
-
-        private readonly string _oldRestoreDirectory;
-        private bool _isRestored;
+        private readonly object _deploymentLock = new object();
+        private IApplicationDeployer _deployer;
+        private DeploymentResult _deploymentResult;
 
         protected ApplicationTestFixture(string applicationName)
         {
             ApplicationName = applicationName;
-            _oldRestoreDirectory = Environment.GetEnvironmentVariable(NuGetPackagesEnvironmentKey);
         }
 
         public string ApplicationName { get; }
 
         public string ApplicationPath => ApplicationPaths.GetTestAppDirectory(ApplicationName);
 
-        public string TempRestoreDirectory { get; } = CreateTempRestoreDirectory();
-
-        public HttpClient HttpClient { get; } = new HttpClient();
+        public HttpClient HttpClient { get; private set; }
 
         public ILogger Logger { get; private set; }
 
-        public IApplicationDeployer CreateDeployment()
+        public DeploymentResult CreateDeployment()
         {
-            PrepareForDeployment();
-            var deploymentParameters = GetDeploymentParameters();
-            return ApplicationDeployerFactory.Create(deploymentParameters, Logger);
-        }
+            lock (_deploymentLock)
+            {
+                if (_deployer != null)
+                {
+                    return _deploymentResult;
+                }
 
-        public virtual void PrepareForDeployment()
-        {
-            Logger = CreateLogger();
+                Logger = CreateLogger();
+                var deploymentParameters = GetDeploymentParameters();
+                var deployer = ApplicationDeployerFactory.Create(deploymentParameters, Logger);
+                _deploymentResult = deployer.Deploy();
 
-            if (!_isRestored)
-            {
-                Restore();
```

Comment:
We currently restore as part of running the tests. This is primarily an artifact of how tools worked with project.json - you couldn't add a project reference to these. We left the behavior as is during the conversion to csproj. Doing this poses a couple of issues

* The tool and the test application might not necessarily reference the same set of versions. This shows up in the test runs in the CI when the apps fail to restore because Coherence recently pushed packages. We could change Universe to pin these apps too. However,
* Running restore requires the tool package to be built and for us to restore to a temp directory to avoid polluting the global packages directory with local builds. This makes the build-test cycle much slower and tests slower yet since it restores the full package graph every time.

I spoke to @Eilon and we decided it was nicer to switch to using a project reference here and use MusicStore as a way to perform E2E testing of the package. Quite a bit of the change here is based on not requiring us to do restore prior to a deployment.